prac_07/project_management.py

```diff
@@ -0,0 +1,172 @@
+            print("Let's add a new project")
+            new_project = create_project()
+            projects.append(new_project)
+        elif menu_choice == "U":
+            for i, project in enumerate(projects):
+                print(f"{i} {project}")
+            project_number_choice = int(input("Project choice: "))
+            print(projects[project_number_choice])
+            projects[project_number_choice] = update_project(projects[project_number_choice])
+        else:
+            print("Invalid choice")
+        print(MENU)
+        menu_choice = input(INPUT_PROMPT).upper()
+    save_to_default_file_choice = input(f"Would you like to save to {DEFAULT_FILENAME}? ").upper()
+    if save_to_default_file_choice == "Y":
+        save_projects(projects, DEFAULT_FILENAME)
+    print("Thank you for using custom-built project management software.")
+
+
+def load_projects(filename):
+    """Load projects from filename, values separated by tabs"""
+    projects = []
+    with open(filename, "r") as in_file:
+        in_file.readline()
+        for line in in_file:
+            parts = line.split("\t")
+            start_date = datetime.datetime.strptime(parts[1], "%d/%m/%Y")
+            project = Project(parts[0], start_date, int(parts[2]), float(parts[3]), int(parts[4]))
+            projects.append(project)
+    print(f"Loaded {len(projects)} projects from {filename}")
+    return projects
+
+
+def save_projects(projects, filename):
+    """Save projects to filename, values separated by tabs"""
+    with open(filename, "w") as out_file:
+        print("Name\tStart Date\tPriority\tCost Estimate\tCompletion Percentage", file=out_file)
+        for project in projects:
+            start_date_string = project.start_date.strftime(DATE_FORMAT_STRING)
+            fields = (project.name, start_date_string, str(project.priority),
+                      str(project.cost_estimate), str(project.percent_complete))
+            row = "\t".join(fields)
+            print(row, file=out_file)
+
+
+def display_projects(projects):
+    """Display projects, sorted by priority and separated based on completion status"""
+    incomplete_projects = sorted([project for project in projects if not project.is_complete()])
+    complete_projects = sorted([project for project in projects if project.is_complete()])
+    print("Incomplete projects:")
+    for incomplete_project in incomplete_projects:
+        print(f"  {incomplete_project}")
+    print("Completed projects:")
+    for complete_project in complete_projects:
+        print(f"  {complete_project}")
+
+
+def filter_projects_by_date(projects):
+    """Display all projects that start on or after a given date"""
+    filter_date = get_valid_date("Show projects that start after date (dd/mm/yyyy): ")
+    filtered_projects = [project for project in projects if project.starts_after_date(filter_date)]
+    for project in sorted(filtered_projects, key=lambda project: project.start_date):
```

Comment:
I think using lambda is more effective, but at JCU Singapore, we are not allowed to use methods that we did not learn in class. How about using attrgetter?

Reply:
Whoops, completely forgot about `attrgetter`. Thanks for the heads up!